server: reduce API discrepancies between SQL-only servers and regular servers

knz opened the pull request:
As of this PR, all the RPCs are left unimplemented (by inheriting from UnimplementedAdminServer).

This also serves the UI (DB Console) on SQL servers, using the same code as regular servers.

Now we can incrementally add more API endpoints as needed.

knz committed:
server: deliver the AdminServer RPC interface on SQL-only servers.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/75852)
<!-- Reviewable:end -->

knz committed:
server: detach authenticationServer from Server, connect in SQL-only

This ensures that the login/logout endpoints become available.

Release note: None

knz committed:
server: detach httpServer from server.Config

... since it does not need a KVConfig.

This moves `EnableDemoLoginEndpoint` and
`EnableWebSessionAuthentication` from KVConfig to BaseConfig, where
they had belonged from the start.

Release note: None

knz committed:
server: pull the /status/vars HTTP handler out of statusServer

Release note: None

knz committed:
server: make `/_status/load` available everywhere

Release note (api change): The `/_status/load` endpoint, which
delivers an instant measurement of CPU load, is now available
for regular CockroachDB nodes and not just multitenant SQL-only
servers.
(This is arguably a bug fix - this endpoint should have been available
from the start.)

knz committed:
server: serve the regular HTTP interface for SQL-only servers

This change reuses the common HTTP server infrastructure for SQL-only
servers, the same one used for regular nodes.

This achieves multiple milestones:

- it makes it possible to serve the DB console on SQL-only servers.
- it makes it possible to navigate SQL sessions and statements
  using a web browser.
- it properly authenticates the `/debug` APIs.

At this point, many parts of the UI are still "blank" (or report
"method not implemented") because the corresponding server-side
support is not yet ... implemented. This will come incrementally.

Release note: None

dhartunian reviewed (commented):
Very exciting!

> Reviewed 3 of 3 files at r1, 5 of 5 files at r2, 4 of 4 files at r3, 3 of 3 files at r4, 6 of 6 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-:-MuvPBOADaZxR0wH-LAv:bvbyvqt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @rimadeodhar, and @srosenberg)

___
*[-- commits, line 16 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-MuvJepSEcprpuK-rrjk):*
nit: maybe mention that this is moving configs from `KVConfig` into `BaseConfig`. Seems like those flags should have been in base anyway.
___
*[pkg/server/server_http.go, line 116 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-MuvK4qX4XswTApB4a52:-MuvK4qX4XswTApB4a53:bt73rz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f68150c1745d235e73c40d526bd51caa7be58a7a/pkg/server/server_http.go#L116)):*
> ```Go
> 	s.mux.Handle(adminHealth, handleRequestsUnauthenticated)
> 	// The /_status/vars endpoint is not authenticated either. Useful for monitoring.
> 	s.mux.Handle(statusVars, http.HandlerFunc(varsHandler{metricSource, s.cfg.Settings}.handleVars))
> ```

I saw this pattern in the httpServer as well that I am modifying in my PR here: https://github.com/cockroachdb/cockroach/pull/72659/files#diff-104a11712684e520f0affa95e6a053faeb713306a10f69ee99b9729cc9dd1996R908

If you make `varsHandler` implement `http.Handler` you no longer need to wrap in `http.HandlerFunc`. Or is there a reason to prefer the wrapping? Since we already have a struct representing the service, it seems natural to have it implement the `Handler` interface.
___
*[pkg/server/tenant.go, line 181 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-MuvNZKW8tlYe7PxeYME:-MuvNZKW8tlYe7PxeYMF:b-ncpkm5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae05f35308655a1670ae4c6b49d1a019910682d2/pkg/server/tenant.go#L181)):*
> ```Go
> 		s.stopper,
> 		grpcMain,
> 		baseCfg.AdvertiseAddr,
> ```

Just to confirm my understanding for this change: all *addresses* should come from`baseCfg` which gets updated when we run `startListenRPCAndSQL` above and the `pgL` listener it returns should not be used to query for addresses.
___
*[pkg/server/tenant.go, line 198 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-MuvOOzeCaEj8voWbLdf:-MuvOOzeCaEj8voWbLdg:b1vfcge) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae05f35308655a1670ae4c6b49d1a019910682d2/pkg/server/tenant.go#L198)):*
> ```Go
> 	httpServer := newHTTPServer(baseCfg)
> 
> 	httpServer.handleHealth(gwMux)
> ```

should we consider creating the server and handling health earlier than here? the non-tenant code seems to call this a lot earlier than `setupRoutes`.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
thanks for the early review! I hope to have the next iteration ready tomorrow.

Do you have some ideas about why TestTenantGRPC is unhappy in CI?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-:-MuvZTlQEGdTk-TMsT7P:bl1vcip)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @knz, @rimadeodhar, and @srosenberg)

___
*[pkg/server/server_http.go, line 116 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-MuvK4qX4XswTApB4a52:-MuvZFC_5sONpF2BDAiy:btwjkzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f68150c1745d235e73c40d526bd51caa7be58a7a/pkg/server/server_http.go#L116)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I saw this pattern in the httpServer as well that I am modifying in my PR here: https://github.com/cockroachdb/cockroach/pull/72659/files#diff-104a11712684e520f0affa95e6a053faeb713306a10f69ee99b9729cc9dd1996R908

If you make `varsHandler` implement `http.Handler` you no longer need to wrap in `http.HandlerFunc`. Or is there a reason to prefer the wrapping? Since we already have a struct representing the service, it seems natural to have it implement the `Handler` interface.
</blockquote></details>

Let's discuss this out of the review and come back to it after the discussion completes.
___
*[pkg/server/tenant.go, line 181 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-MuvNZKW8tlYe7PxeYME:-MuvZJEH5xzM4iba6MUh:b2nign6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae05f35308655a1670ae4c6b49d1a019910682d2/pkg/server/tenant.go#L181)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Just to confirm my understanding for this change: all *addresses* should come from`baseCfg` which gets updated when we run `startListenRPCAndSQL` above and the `pgL` listener it returns should not be used to query for addresses.
</blockquote></details>

yes


<!-- Sent from Reviewable.io -->

rimadeodhar reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75852)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @knz, and @srosenberg)

___
*[pkg/server/tenant_admin.go, line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-Muw_wmsCSZKzQf9u59G:-Muw_swPDCR5z-7Wu2yV:b-wea9el) ([raw file](https://github.com/cockroachdb/cockroach/blob/080e4345296092e19ca3cdbcd4301bbb0fff6624/pkg/server/tenant_admin.go#L55)):*
> ```Go
> 
> // TODO(knz): add Drain implementation here.
> // TODO(rima): add Nodes implementation here for debug zip.
> ```

I'm confused about why we would want to add Nodes implementation to the `tenantAdminServer` as opposed to the `tenantStatusServer`? It is a part of the statusClient currently. Also, would we want to implement the `Nodes` API at all for SQL only servers given that the NodesResponse object contains a fair number of fields that don't have much meaning in the SQL only server context? If it is mainly about having parity between the tenantStatusServer and statusServer then that makes sense but the usability of the Nodes API would be fairly restricted for the SQL only servers


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75852)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @knz, @rimadeodhar, and @srosenberg)

___
*[pkg/server/tenant.go, line 198 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-MuvOOzeCaEj8voWbLdf:-Mv328z2-e0brIzdiZT0:b-7vl4mn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae05f35308655a1670ae4c6b49d1a019910682d2/pkg/server/tenant.go#L198)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

should we consider creating the server and handling health earlier than here? the non-tenant code seems to call this a lot earlier than `setupRoutes`.
</blockquote></details>

Yes that would perhaps be a good idea, at least after we split the constructor code from the "start" code.
(As is done for non-tenant code, but not currently the case here.)

However, that would be a different type of change, which I believe is beyond the scope of the PR here. I wanted to keep this one neutral wrt initialization order so as to limit the amount of code movement.
___
*[pkg/server/tenant_admin.go, line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-Muw_wmsCSZKzQf9u59G:-Mv32T6w0gpSeJTOOsjK:b-52o3vi) ([raw file](https://github.com/cockroachdb/cockroach/blob/080e4345296092e19ca3cdbcd4301bbb0fff6624/pkg/server/tenant_admin.go#L55)):*
> It is a part of the statusClient currently.

Oh my, you are right. I got myself confused too. Removed this comment.

>  Also, would we want to implement the Nodes API at all for SQL only servers given that the NodesResponse object contains a fair number of fields that don't have much meaning in the SQL only server context?

What you did in the other PR, adding a new endpoint, is fine. However, I can imagine a transition period, where we increase coverage of the existing API to bridge compatibility with the existing UI code (which uses Nodes), and only seek to remove/switch APIs in a second phase.


<!-- Sent from Reviewable.io -->

knz commented:
@maryliag @dhartunian can you have folk in your team review the test API changes and the unit test changes in ccl/serverccl.
I think it's important for folk to understand that *unit tests too need to use authentication to use HTTP APIs*.

(And conversely, we need to train every member of every team to understand that any network API needs authentication + authorization checks. It's unfortunate that unit tests get written and approved during reviews without checks for authn/authz.)

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75852)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @rimadeodhar, and @srosenberg)

___
*[-- commits, line 16 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-MuvJepSEcprpuK-rrjk):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: maybe mention that this is moving configs from `KVConfig` into `BaseConfig`. Seems like those flags should have been in base anyway.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz committed:
server: move the contention registry instantiation out of server.go

... and tenant.go, into `newSQLServer()` where it belongs.

Release note: None

knz reviewed (commented):
> Do you have some ideas about why TestTenantGRPC is unhappy in CI?

This was the missing authentication. Fixed it.

The PR is now ready for review. Thanks

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-:-Mv3ljuq0CT5ZvReQqAB:b-wa6bz8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @rimadeodhar, and @srosenberg)



<!-- Sent from Reviewable.io -->

knz committed:
server: move httpTestServer out of testserver.go

Release note: None

knz reviewed (commented):
If so desired, I can break it up in smaller PRs, one per commit?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-:-Mv3mg3PA-_6bqYyRO9o:bjbpaod)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @rimadeodhar, and @srosenberg)



<!-- Sent from Reviewable.io -->

maryliag reviewed (approved):
Regarding the unit test changes in ccl/serverccl: <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 6 files at r14, 1 of 4 files at r15, 1 of 3 files at r16, 1 of 6 files at r17, 6 of 11 files at r18, 3 of 3 files at r19, 3 of 3 files at r20, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-:-MvOn6lC3n6UxrIK-Ot7:bmklwcj)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dhartunian, @rimadeodhar, and @srosenberg)



<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> just a few small questions

> (And conversely, we need to train every member of every team to understand that any network API needs authentication + authorization checks. It's unfortunate that unit tests get written and approved during reviews without checks for authn/authz.)

Regarding this issue, to what extent can this be nudged by making auth the default mode? Renaming `GetHTTPClient` to `GetUnauthenticatedHTTPClient()` could be enough to help people have awareness that security boundaries are being crossed.

> Reviewed 17 of 17 files at r13, 6 of 6 files at r14, 22 of 22 files at r21, 5 of 5 files at r22, 4 of 4 files at r23, 3 of 3 files at r24, 6 of 6 files at r25, 11 of 11 files at r26, 3 of 3 files at r27, 3 of 3 files at r28, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-:-MvPAd7fCee0fUEe97OS:br9j940)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @knz, @rimadeodhar, and @srosenberg)

___
*[pkg/server/testserver.go, line 844 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-MvPFIiE6FFlJK_LkDFq:-MvPFIiE6FFlJK_LkDFr:br0zp8h) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d010f7bbf63b4f5ed41b191abf225c31deb641c/pkg/server/testserver.go#L844)):*
> ```Go
> }
> 
> // AdminURL implements TestServerInterface.
> ```

Do we need to take off these comments?
___
*[pkg/testutils/serverutils/test_server_shim.go, line 143 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-MvPFUFwAVPnTgPbrdSo:-MvPFUFwAVPnTgPbrdSp:b-216ca5) ([raw file](https://github.com/cockroachdb/cockroach/blob/498970cac65033d78529fd329b60cc0b76afe139/pkg/testutils/serverutils/test_server_shim.go#L143)):*
> ```Go
> 	// GetAuthenticatedHTTPClient returns an http client which has been
> 	// authenticated to access Admin API methods (via a cookie).
> 	GetAuthenticatedHTTPClient(isAdmin bool) (http.Client, error)
> ```

I don't quite follow why these methods get removed from `TestServerInterface` but get added to `TestTenantInterface`.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> Renaming GetHTTPClient to GetUnauthenticatedHTTPClient() could be enough to help people have awareness that security boundaries are being crossed.

Ooh this is a _great_ idea. Let's do this. Filed as #76241

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-:-MvPI2qU3b984Tx_TNBj:bm9hfjx)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @dhartunian, @rimadeodhar, and @srosenberg)

___
*[pkg/server/testserver.go, line 844 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-MvPFIiE6FFlJK_LkDFq:-MvPIgrVD9ZxBCoS9R7t:bkpzvp) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d010f7bbf63b4f5ed41b191abf225c31deb641c/pkg/server/testserver.go#L844)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Do we need to take off these comments?
</blockquote></details>

Which comments? I think they are removed in the final commit.
___
*[pkg/testutils/serverutils/test_server_shim.go, line 143 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/75852#-MvPFUFwAVPnTgPbrdSo:-MvPIjaQAiV1p1C0Af24:bjpfg6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/498970cac65033d78529fd329b60cc0b76afe139/pkg/testutils/serverutils/test_server_shim.go#L143)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I don't quite follow why these methods get removed from `TestServerInterface` but get added to `TestTenantInterface`.
</blockquote></details>

TestServerInterface embeds TestTenantInterface.

Maybe we need to rename these two interfaces (maybe a different PR):
- TestTenantInterface contains the *common* APIs, valid for both sql-only and mixed servers.
- TestServerInterface contains the APIs that are only useful for mixed KV/SQL servers.


<!-- Sent from Reviewable.io -->

knz commented:
I'm going to go ahead and merge this, and then we can follow-up with more PRs as needed for more polish.

bors r=dhartunian,rimadheodar,maryliag
